File: 06-Lanternfish.hs
import Data.List.Split (splitOn)
import Data.List (group,sort)

-- runhaskell 06-Lanternfish.hs <06-input.txt

toi :: String -> Int
toi s = read s::Int

-- | Translate a list of fish states into a list of number of fishes per state
initState :: [Int] -> [Int]
initState fish = map (\xs -> (length xs - 1)) $ group $ sort (fish ++ [0..8])

-- | Calculate one day iteration
nextState :: [Int] -> [Int]
nextState state = shifted ++ [breeders]
  where
    breeders = head state
    rest = tail state
    shifted = (take 6 rest) ++ [( rest !! 6 ) + breeders] ++ (drop 7 rest)


-- | Calculate the next state over a given amount of days
advanceTime :: Int -> [Int] -> [Int]
advanceTime days state = foldl (\state _ -> (nextState state)) state [1..days]

-- | Sum list
countFishes :: [Int] -> Int
countFishes = foldl (+) 0

main = do
  contents <- getContents
  let input = map toi $ head $ map (splitOn ",") $ lines contents
      initialState = initState input

  print $ countFishes $ advanceTime 80 initialState
  print $ countFishes $ advanceTime 265 initialState

File: 2021/06-Lanternfish.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

let input

rl.on("line", data => {
 input =  data.split(',').map(i=>Number.parseInt(i))

})

rl.on('close', () => {
  console.log('partOne', solve(input, 80))  // 5934        // 387413
  console.log('partTwo', solve(input, 256)) // 26984457539 // 1738377086345
})


function solve(input, days=10) {
  let state = input.reduce(
    (state, i) => {state[i]++; return state},
  Array(9).fill(0))


  for (let d=0; d<days; d+=1) {
    const [breeders, ...rest] = state
    state = [...rest, breeders]
    state[6]+=breeders
  }

  return state.reduce((a,b)=>a+b)
} 

